File: backend/models/userModel.js
import mongoose from "mongoose";
import bcrypt from "bcryptjs";

const userSchema = new mongoose.Schema(
    {
        // ===================================
        // INFORMACIÓN PERSONAL
        // ===================================

        name: {
            type: String,
            required: true,
            trim: true
        },

        lastName: {
            type: String,
            trim: true,
            default: ""
        },

        email: {
            type: String,
            required: true,
            lowercase: true,
            trim: true,
            unique: true
        },

        phone: {
            type: String,
            trim: true,
            default: ""
        },

        // ===================================
        // AUTENTICACIÓN TRADICIONAL
        // ===================================

        // La contraseña no es obligatoria porque
        // los usuarios autenticados mediante Google
        // pueden no tener contraseña.
        password: {
            type: String,
            minlength: 6,
            maxlength: 100,
            default: null
        },

        // ===================================
        // AUTENTICACIÓN CON GOOGLE
        // ===================================

        googleId: {
            type: String,
            unique: true,
            sparse: true,
            default: undefined
        },

        // ===================================
        // IMAGEN DE PERFIL
        // ===================================

        profileImage: {
            type: String,
            default: ""
        },

        // ===================================
        // INFORMACIÓN DEL USUARIO
        // ===================================

        role: {
            type: String,
            enum: ["ADMIN", "USER"],
            default: "USER"
        },

        points: {
            type: Number,
            default: 0,
            min: 0
        },

        // Estado general de la cuenta.
        // Permite activar o desactivar usuarios
        // desde la administración.
        isActive: {
            type: Boolean,
            default: true
        },

        // Estado de verificación del correo.
        // Los usuarios tradicionales deben verificar
        // su correo antes de iniciar sesión.
        isEmailVerified: {
            type: Boolean,
            default: false
        },

        // ===================================
        // RECUPERACIÓN DE CONTRASEÑA
        // ===================================

        recoveryCode: {
            type: String,
            default: null
        },

        recoveryCodeExpiration: {
            type: Date,
            default: null
        }
    },
    {
        timestamps: true,
        collection: "users"
    }
);

// ===================================
// HASH DE CONTRASEÑA
// ===================================

userSchema.pre("save", async function () {

    // No intentar encriptar una contraseña
    // que no existe.
    if (!this.password) {
        return;
    }

    // No volver a encriptar una contraseña
    // que ya fue almacenada.
    if (!this.isModified("password")) {
        return;
    }

    // Generar salt.
    const salt = await bcrypt.genSalt(10);

    // Encriptar contraseña.
    this.password = await bcrypt.hash(
        this.password,
        salt
    );
});

// ===================================
// USER MODEL
// ===================================

const User = mongoose.model(
    "User",
    userSchema
);

export default User;

File: frontend/lib/models/user_model.dart
class AppUser {
  final String id;
  final String name;
  final String lastName;
  final String email;
  final String role;
  final String phone;
  final int points;
  final String profileImage;
  final bool isActive;
  final bool isEmailVerified;

  AppUser({
    required this.id,
    required this.name,
    required this.lastName,
    required this.email,
    required this.role,
    required this.phone,
    required this.points,
    required this.profileImage,
    required this.isActive,
    required this.isEmailVerified,
  });

  factory AppUser.fromJson(Map<String, dynamic> json) {
    return AppUser(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      name: json['name'] as String? ?? '',
      lastName: json['lastName'] as String? ?? '',
      email: json['email'] as String? ?? '',
      role: json['role'] as String? ?? 'USER',
      phone: json['phone'] as String? ?? '',
      points: (json['points'] as num?)?.toInt() ?? 0,
      profileImage: json['profileImage'] as String? ?? '',
      isActive: json['isActive'] as bool? ?? true,
      isEmailVerified: json['isEmailVerified'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'lastName': lastName,
        'email': email,
        'role': role,
        'phone': phone,
        'points': points,
        'profileImage': profileImage,
        'isActive': isActive,
        'isEmailVerified': isEmailVerified,
      };
}
